File: ts/src/asana-reader.ts
/** Read tasks from Asana via the official Node.js SDK (v3). */

import * as asana from "asana";
import { buildTaskData } from "./taskwarrior.js";
import type { Config, TaskData } from "./types.js";

const OPT_FIELDS = [
  "gid",
  "name",
  "completed",
  "due_on",
  "tags",
  "tags.name",
  "projects",
  "projects.name",
  "notes",
];

const SUBTASK_OPT_FIELDS = "gid,name,completed,due_on,notes";
const STORY_OPT_FIELDS = "text,created_by,created_by.name,type,created_at";

async function getWorkspaceGid(
  workspacesApi: asana.WorkspacesApi,
  configured: string,
): Promise<string> {
  if (configured) return configured;

  const response = await workspacesApi.getWorkspaces({});
  const workspaces = response.data ?? [];

  if (workspaces.length === 0) {
    throw new Error("No Asana workspaces found for this account.");
  }
  if (workspaces.length > 1) {
    const names = workspaces.map((w: { name?: string }) => w.name).join(", ");
    throw new Error(
      `Multiple workspaces found (${names}). ` +
        "Set 'workspace' in your config to pick one.",
    );
  }
  return workspaces[0].gid;
}

export async function fetchAsanaTasks(config: Config): Promise<TaskData[]> {
  const asanaCfg = config.asana;
  const syncCfg = config.sync;
  const sourceTag = syncCfg.asana_tag;

  const token = asanaCfg.personal_access_token;
  if (!token || token === "YOUR_TOKEN_HERE") {
    throw new Error(
      "Asana personal access token not configured. " +
        "Run 'todo-sync-ts setup' or edit ~/.config/todo-sync/config.toml",
    );
  }

  // Set up the API client with authentication
  const client = new asana.ApiClient();
  client.authentications["token"].accessToken = token;

  const workspacesApi = new asana.WorkspacesApi(client);
  const userTaskListsApi = new asana.UserTaskListsApi(client);
  const tasksApi = new asana.TasksApi(client);
  const storiesApi = new asana.StoriesApi(client);

  const workspaceGid = await getWorkspaceGid(
    workspacesApi,
    asanaCfg.workspace ?? "",
  );

  // Get user's task list
  const userTaskList = await userTaskListsApi.getUserTaskListForUser(
    "me",
    workspaceGid,
  );

  // Fetch tasks from the user's task list (only incomplete tasks)
  const rawTasks: Array<{
    gid: string;
    name: string;
    completed: boolean;
    due_on: string | null;
    tags: Array<{ name?: string }> | null;
    projects: Array<{ name?: string }> | null;
    notes: string | null;
  }> = [];

  let offset: string | undefined;
  do {
    const opts: Record<string, unknown> = {
      opt_fields: OPT_FIELDS.join(","),
      limit: 100,
      completed_since: "now", // Only incomplete tasks
    };
    if (offset) opts.offset = offset;

    const tasksResponse = await tasksApi.getTasksForUserTaskList(
      userTaskList.data.gid,
      opts,
    );
    rawTasks.push(...(tasksResponse.data ?? []));
    offset = tasksResponse._response?.next_page?.offset;
  } while (offset);

  const result: TaskData[] = [];
  const allSubtasks: TaskData[] = [];

  for (let i = 0; i < rawTasks.length; i++) {
    const t = rawTasks[i];

    // Skip if somehow completed (shouldn't happen with completed_since: 'now')
    if (t.completed) continue;

    // Log progress every 10 tasks
    if (i % 10 === 0) {
      console.log(`  Fetching details for task ${i + 1}/${rawTasks.length}...`);
    }

    // Collect tag names
    const tags: string[] = [];
    if (t.tags) {
      for (const tag of t.tags) {
        if (tag.name) tags.push(tag.name);
      }
    }

    // Get first project name
    let projectName: string | undefined;
    if (t.projects) {
      for (const p of t.projects) {
        if (p.name) {
          projectName = p.name;
          break;
        }
      }
    }

    // Build annotations from notes
    const annotations: string[] = [];
    if (t.notes) annotations.push(t.notes);

    // Fetch comments (stories) for this task
    try {
      const storiesResponse = await storiesApi.getStoriesForTask(t.gid, {
        opt_fields: STORY_OPT_FIELDS,
      });
      const stories = storiesResponse.data ?? [];
      for (const story of stories) {
        if (story.type === "comment" && story.text) {
          const author = story.created_by?.name ?? "Unknown";
          const commentText = `[${author}] ${story.text}`.slice(0, 1000);
          annotations.push(commentText);
        }
      }
    } catch {
      // Non-fatal: continue without comments
    }

    const taskData = buildTaskData({
      description: t.name,
      externalIdField: "asana_gid",
      externalId: t.gid,
      source: "asana",
      sourceTag,
      project: projectName,
      due: t.due_on ?? undefined,
      tags,
      annotations,
    });

    result.push(taskData);

    // Fetch subtasks for this task
    try {
      const subtasksResponse = await tasksApi.getSubtasksForTask(t.gid, {
        opt_fields: SUBTASK_OPT_FIELDS,
      });
      const subtasks = subtasksResponse.data ?? [];
      for (const sub of subtasks) {
        if (sub.completed) continue;

        const subTaskData = buildTaskData({
          description: sub.name,
          externalIdField: "asana_gid",
          externalId: sub.gid,
          source: "asana",
          sourceTag,
          project: projectName,
          due: sub.due_on ?? undefined,
          tags,
          annotations: sub.notes ? [sub.notes] : undefined,
        });
        subTaskData.parentAsanaGid = t.gid;
        allSubtasks.push(subTaskData);
      }
    } catch {
      // Non-fatal: continue without subtasks
    }
  }

  return [...result, ...allSubtasks];
}

File: src/todo_sync/asana_reader.py
"""Read tasks from Asana via the official Python SDK."""

from __future__ import annotations

from datetime import datetime

import asana
from asana.rest import ApiException

from todo_sync.taskwarrior import build_task_data

# Fields to request from the Asana API
OPT_FIELDS = "gid,name,completed,due_on,tags,tags.name,projects,projects.name,notes"


def _get_workspace_gid(api_client: asana.ApiClient, configured: str) -> str:
    """Resolve workspace GID: use configured value or auto-detect."""
    if configured:
        return configured

    workspaces_api = asana.WorkspacesApi(api_client)
    workspaces = list(workspaces_api.get_workspaces(opts={"limit": 10}))

    if not workspaces:
        raise RuntimeError("No Asana workspaces found for this account.")
    if len(workspaces) > 1:
        names = ", ".join(w.name for w in workspaces)
        raise RuntimeError(
            f"Multiple workspaces found ({names}). "
            "Set 'workspace' in your config to pick one."
        )
    return workspaces[0].gid


def fetch_asana_tasks(config: dict) -> list[dict]:
    """Fetch incomplete tasks from Asana and return normalized task dicts."""
    asana_cfg = config.get("asana", {})
    sync_cfg = config.get("sync", {})
    source_tag = sync_cfg.get("asana_tag", "asana")

    token = asana_cfg.get("personal_access_token", "")
    if not token or token == "YOUR_TOKEN_HERE":
        raise RuntimeError(
            "Asana personal access token not configured. "
            "Run 'todo-sync setup' or edit ~/.config/todo-sync/config.toml"
        )

    configuration = asana.Configuration()
    configuration.access_token = token
    api_client = asana.ApiClient(configuration)

    workspace_gid = _get_workspace_gid(api_client, asana_cfg.get("workspace", ""))

    # Get user's task list
    user_task_lists_api = asana.UserTaskListsApi(api_client)
    user_task_list = user_task_lists_api.get_user_task_list_for_user(
        user_gid="me",
        workspace=workspace_gid,
        opts={},
    )

    # Fetch tasks from the user's task list
    tasks_api = asana.TasksApi(api_client)
    raw_tasks = tasks_api.get_tasks_for_user_task_list(
        user_task_list_gid=user_task_list.gid,
        opts={"opt_fields": OPT_FIELDS},
    )

    result = []
    for t in raw_tasks:
        # Skip completed tasks
        if t.completed:
            continue

        # Parse due date
        due = None
        if t.due_on:
            try:
                due = datetime.strptime(t.due_on, "%Y-%m-%d")
            except (ValueError, TypeError):
                pass

        # Collect tag names
        tags = []
        if t.tags:
            tags = [tag.name for tag in t.tags if hasattr(tag, "name")]

        # Get first project name
        project_name = None
        if t.projects:
            for p in t.projects:
                if hasattr(p, "name") and p.name:
                    project_name = p.name
                    break

        # Build annotations from notes
        annotations = []
        if t.notes:
            annotations.append(t.notes)

        task_data = build_task_data(
            description=t.name,
            external_id_field="asana_gid",
            external_id=t.gid,
            source="asana",
            source_tag=source_tag,
            project=project_name,
            due=due,
            tags=tags,
            annotations=annotations,
        )
        result.append(task_data)

    return result
